Thread/process leak issue fixes

GregoryMorse opened the pull request:
Fix thread handle leak of decompilation thread in successful decompilation cases

GregoryMorse committed:
Update idaplugin.cpp

Fix thread handle leak of decompilation thread in successful decompilation cases

GregoryMorse committed:
Update defs.h

Track process handle

GregoryMorse committed:
Update defs.cpp

Fix Windows thread handle leak and save process handle to avoid leaking that and have better control over the process status and exit or termination

GregoryMorse committed:
Update decompiler.cpp

Process handle save

GregoryMorse committed:
Update idaplugin.cpp

Terminate process by handle which should close the handle

GregoryMorse committed:
Update defs.cpp

Portable handle saving for Linux/Windows

GregoryMorse commented:
The real question is whether or not term_process in IDA closes the process handle as well as check_process_exit when an exit actually occurs.  If it does not then this reduces the portability of IDAs interface and seems like bad behavior.

The taskkill/kill method certain should leak the handle.  If IDA is using PIDs as the void pointer returned then this is a short coming in IDA and on the Windows side conditional compilation should uses the hProcess for WaitForSingleObject or TerminateProcess and CloseHandle.  If p == pi.hProcess inside of runCommand which it should, then probably IDA cleans it up on the check or term call.  If not then either the handle was duplicated which would be unlikely and strange and require explicit Windows closing, or its using PIDs again which can cause all sorts of race conditions or unexpected situations on a high load system given the lack of a recycling guarantee by the OS.

The current changes proposed regardless should harden against leaks if IDA has expected behavior.

GregoryMorse committed:
Update idaplugin.cpp

GregoryMorse commented:
This took only process explorer and single WARNING_GUI extra debug trace.

p == pi.hProcess as expected and indeed term_process and check_process_exit seem to manage this handle.

But surely not calling thread.join() leaks a handle every run as does not calling CloseHandle on pi.hThread - pretty strange behavior but I suppose this extra handle was not intended to be managed.  The docs should mention it.

The fix stands as is in its current version.

s3rvac commented:
Thank you for taking the time to analyze the issue and submit a PR :+1: @PeterMatula will review it.

GregoryMorse committed:
Update code_viewer.cpp

Remove thread leak and unnecesary main thread -> thread -> main thread complication (only decompile.cpp invocation comes from another thread).

GregoryMorse committed:
Update idaplugin.cpp

Remove thread leak and unnecesary main thread -> thread -> main thread complication (only decompile.cpp invocation comes from another thread).

GregoryMorse committed:
Update idaplugin.cpp

GregoryMorse commented:
The change is ready to go - this satisfies my analysis - there is some edge case in the issue I discussed on plugin termination if for some reason the thread tries to execute work on the main thread at that exact moment - so the thread kill would have to suffice there, it is good enough.  There is complicated code with a semaphore and mutex to work around it but I think such an extremely unlikely scenario is not necessarily worth over-complicating the project for now.

s3rvac commented:
Thank you for the changes. @PeterMatula, can you please review the code?

PeterMatula commented on src/idaplugin/code_viewer.cpp:
```diff
@@ -357,7 +357,8 @@ bool idaapi moveToPrevious()
 
 		decompInfo.decompiledFunction = fit->first;
-		qthread_create(showDecompiledCode, static_cast<void*>(&decompInfo));
+		ShowOutput show(&decompInfo);
```
You are right about this being a remnant of an old design. Creating thread here (and other 3 places in this source and 1 place in `idaplugin.cpp`) is unnecessary.

PeterMatula commented on src/idaplugin/defs.cpp:
```diff
@@ -20,7 +20,8 @@ int runCommand(
 		intptr_t* pid,
-		bool showWarnings)
+		bool showWarnings,
+		void** hdl)
```
Is it possible that you forgot to commit the corresponding change in function declaration in `defs.h`? I'm unable to compile this, because `defs.h` does not have `void** hdl` argument. It works when I do this in `defs.h` (note the default value, it is important for the next comment):
```c
int runCommand(
		const std::string& cmd,
		const std::string& args,
		intptr_t* pid = nullptr,
		bool showWarnings = false,
		void** hdl = nullptr);
```
What is your code? Do you have something different in there?

GregoryMorse replied:
Yes it is exactly right, I forgot to update the header definition.  Thanks for finding commit it now.

PeterMatula commented on src/idaplugin/defs.cpp:
```diff
@@ -44,6 +45,10 @@ int runCommand(
 		}
 		return 1;
 	}
+	*hdl = p;
```
How is it possible that this does not crash for you? `runCommand()` is also used in `RdGlobalInfo::initPythonCommand()` in `defs.cpp` module - this time with default parameters - i.e. `hdl` is not set. Setting it here when `hdl` is `nullptr` crashes for me. I think there should be something like this:
```c
if (hdl)
{
	*hdl = p;
}
```
Or the `hdl` can not have a default value set and `runCommand()` uses in `RdGlobalInfo::initPythonCommand()` need to change.

You likely have Python interpreter explicitly set so that the code causing problems in `RdGlobalInfo::initPythonCommand()` does not get executed?

GregoryMorse replied:
Yes I did not test it when the pointer was null - but absolutely it needs to be guarded.  Thanks for this as well.  And yes I was not using the same Python init.

PeterMatula commented on src/idaplugin/idaplugin.cpp:
```diff
@@ -40,16 +40,23 @@ void killDecompilation()
+			int rc = 0;
+			if (check_process_exit(decompInfo.hDecomp, &rc, 0) != 0)
+				term_process(decompInfo.hDecomp);
+/*#if defined(OS_WINDOWS)
```
Well, I guess this is the main issue of this issue :D
I was not aware that the current solution is leaking, but I must admit I wasn't really checking. The problem I tried to solve is to kill not only the spawned process, but the whole process tree. I experimented with IDA SDK's `check_process_exit()` and `term_process()`, but it looks like it terminates only the originally started process - i.e. `retdec-decompiler.py` script. The big problem is, that this script spawns other processes (fortunately only one at the time). So if we kill it from IDA using SDK function, processes like `retdec-bin2llvmir` and `retdec-llvmir2hll` keep running. And they can take a very long time on some inputs. And if user triggers several such long decompilations (and kills them), they stack.

This is caused by RetDec's design (it is not possible to used it like a library at the moment) and a whole IDA plugin is kind of a hack that tries to keep it together.

It would be great, if we could come up with a solution that solves both problems: 1) kills an entire process tree, and 2) does not leak. It looks like you have quite an insight into these things (more than I do), so maybe you will know the answer. But right now, I don't think it would be a step forward to merge this solution solving leaks, but leaving potentially endless processes running on user machines.

GregoryMorse replied:
I did not realize about the process tree - yes now I understand why you have the special OS-specific code in there.

As far as I know, the process tree will be terminated automatically only if a job is created.  CreateJobObject / AssignProcessToJobObject / SetInformationJobObject with JOB_OBJECT_LIMIT_KILL_ON_JOB_CLOSE / TerminateJobObject.  Other hack like solutions exist like DEBUG_PROCESS and DEBUG_ONLY_THIS_PROCESS WaitForDebugEvent/ContinueDebugEvent taking advantage of a side effect.

A nice solution also is to make the child processes use stdin pipe and let the pipe closing automatically terminate but of course this works only for an IO driven process.

PeterMatula reviewed (changes requested):
Thanks for this. I was not aware of the leaks. Some changes are OK and could be merged as they are. However, the main issue in `killDecompilation()` is more compilated and your solution have some drawbacks. I hope we can come up with a definitive solution. We can discuss this further in the pull request comments.

GregoryMorse committed:
Update defs.h

GregoryMorse committed:
Update defs.cpp

GregoryMorse commented:
After fixing the 2 issues pointed out by @PeterMatula , it should compile, run without issue.  The process tree termination code may need to be backed out as its a non-trivial situation if the child does not auto detect with pipes (overhead of jobs on Windows and Linux also has its own solution).  There is always the child process finding possibility using special process enumeration API in Windows as well - Process32First/Process32Next from toolhelp or even the low level NtQueryProcessInformation among others.

Of course killing the thread is completely unsafe and will definitely cause leaks.  Better to kill the process and let the thread terminate for some time e.g. 1500ms, then kill as a total last resort.  The thread check_process_exit will close the process handle.  Otherwise if the thread is killed and process killed that should be called from the main thread to close the process handle (since term_process is not an option).  This is a much cleaner termination strategy - since probably few users want to have IDA running after a thread kill due to risks of database corruption.

As an aside, the runCommand's for python do not require the handle.  They are run on the main thread and blocking calls so no issue there.  I think it all looks clean now.

GregoryMorse commented:
I will post my termination fix shortly so this PR can be mergable.  My hope is the plugin will be quite resilient with these changes - in terms of loading/unloading, processes and threads, handles.

GregoryMorse committed:
Update idaplugin.cpp

PeterMatula commented:
Thank you very much for this, I would have hard time trying to figure this out.

I made some minor cosmetic fixes after the merge in 50d63e3.

I'm also looking at `decompRunning` and thinking if it is even necessary. What do you think about `killDecompilation()` looking a bit simpler?
```c++
void killDecompilation()
{
	if (decompInfo.decompPid)
	{
		INFO_MSG("Unfinished decompilation was KILLED !!! Only one decompiltion can run at a time.\n");

		// The following is not used because entire process tree must be terminated.
		//int rc = 0;
		//if (check_process_exit(decompInfo.hDecomp, &rc, 0) != 0)
		//	term_process(decompInfo.hDecomp);
#if defined(OS_WINDOWS)
		std::string cmd = "taskkill /F /T /PID " + std::to_string(decompInfo.decompPid);
		std::system(cmd.c_str());
#else // Linux || macOS
		kill(decompInfo.decompPid, SIGTERM);
#endif

		// This is probably too risky without user knowing. IDA could be unstable.
		// The process kill being successful above, this should be unnecessary.
		//qthread_kill(decompInfo.decompThread);

		qthread_join(decompInfo.decompThread);
		qthread_free(decompInfo.decompThread);
		decompInfo.decompPid = 0;
	}
}
```

GregoryMorse commented:
Right now decompRunning is needed for cleaning up the thread - the thread cannot clean itself up and so its always cleaned up on next UI event be it a run or term.  This way we do not kill the process tree if the thread exited.  Of course the Pid being set might be an unreliable determiner of the thread having run - perhaps the process launch failed.  I guess the correct behavior is to keep decompRunning, though admittedly hDecomp which I added is no longer necessary I now realize, basically just a pointless variable (left it because it can be used to stop process handle leaking of qthread_kill would be executed).  If the pid were used to decide to kill process, and the thread handle itself could be checked for joining and freeing the thread that would be adequate.  However, the decision as to whether the thread is still running and to prevent another try would be handled.  The thread could get launched potentially twice at the same time as there is a race condition since the process could terminate yet the thread has not yet exited.  Only decompRunning guards against it, a pretty small point and maybe irrelevant since the threads call back the UI thread for most of the work - but the data accesses in threads are not thread-safe.  So for this reason the current code is still good.  I could add code to use toolhelp Process32First/Next to kill the child process tree more elegantly than using the shell but its just adding an ugly blob of Win32 code which does exactly the subset of what taskkill is doing, a utility seeming to be guaranteed on all the support Windows versions.  The benefit however, is that hDecomp could be used to wait/term_process after it and its children are killed with TerminateProcess.  Presumably the thread kill would be okay then considering is basically assured to be in the wait code - dynamic memory would also need to be removed from the entire local variables on the whole stack up to that point.

How unsafe is a thread kill?  Well it really depends what the thread is doing - besides leaking the process handle (most likely since that's where its waiting unless a total fluke occurs), it could leak dynamic memory especially when STL classes are used or even other handles like the thread handle or file handles.  I would not exactly deem that destabilizing but certainly its best to avoid except in threads whose handle/memory allocations are completely and safely tracked outside of it probably meaning no kill without checking synchronization objects etc.

You are welcome :).  Most people do not care about these low level OS details - I wanted to share this to enhance the stability of the plugin, while saving the main team time to work on the core.  The speed of the context menu actions seems to be quite a bit more fast and responsive - the most noticeable part of the fix beyond the mere subtlety of the leaking which only becomes a real issue with aggressive successive usage.
